azure-core/azure/core/pipeline/policies/custom_hook.py

```diff
@@ -0,0 +1,46 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+"""
+This module is the requests implementation of Pipeline ABC
+"""
+from .base import SansIOHTTPPolicy
+
+class CustomHookPolicy(SansIOHTTPPolicy):
+    """A simple policy that enable the given callback
+    with the response.
+    """
+    def __init__(self, **kwargs):
+        self._cls = None
+    
+    def on_request(self, request):
+        # type: (PipelineRequest) -> None
+        self._cls = request.context.options.pop('raw_response_hook', {})
```

Comment:
Are we `pop`:ing all kwargs in other policies? 

High-order bit, I expect all policies to be consistent in how they deal with kwargs. 

Low-order bit, should policies do a `get` rather than a `pop`? @annatisch, @bryevdv, thoughts?

Reply:
In the transport, we do not allow pre-defined keys. So here we need to pop the key and then add it back in on_response

Reply:
I definitely prefer it when kwargs for distinct base classes are disjoint, and in that case I prefer to `pop` the relevant kwargs off in each initializer. This affords the possibility of detecting extra/unused kwargs at a top level base and issuing a warning/error. If there are kwargs re-used between different bases this is not possible to do cleanly. Overall I agree that classes should be mutually consistent, whatever they do.

Reply:
Tho I guess this situation is a bit different than the cooperative inheritance case. I don't have a strong opinion, if the downstream demands fixed keys, then those have to be supplied. I don't understand the default value for `pop` tho. Shouldn't it be something like `lambda x: x`, i.e an identity function?

Reply:
Changed to None.